drools-test-coverage/test-compiler-integration/src/test/java/org/drools/compiler/integrationtests/ConstraintWithAndOrConcurrencyJittingTest.java

```diff
@@ -0,0 +1,238 @@
+import java.util.concurrent.TimeUnit;
+
+import org.drools.mvel.expr.MvelEvaluator;
+import org.drools.testcoverage.common.util.KieBaseTestConfiguration;
+import org.drools.testcoverage.common.util.KieBaseUtil;
+import org.drools.testcoverage.common.util.TestParametersUtil;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.kie.api.KieBase;
+import org.kie.api.runtime.KieSession;
+import org.kie.test.testcategory.TurtleTestCategory;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(Parameterized.class)
+@Category(TurtleTestCategory.class)
+public class ConstraintWithAndOrConcurrencyJittingTest {
+
+    private static int LOOP = 1000;
+
+    private static int THREADS = 32;
+    private static int REQUESTS = 32;
+
+    private final KieBaseTestConfiguration kieBaseTestConfiguration;
+
+    public ConstraintWithAndOrConcurrencyJittingTest(final KieBaseTestConfiguration kieBaseTestConfiguration) {
+        this.kieBaseTestConfiguration = kieBaseTestConfiguration;
+    }
+
+    @Parameterized.Parameters(name = "KieBase type={0}")
+    public static Collection<Object[]> getParameters() {
+        return TestParametersUtil.getKieBaseCloudConfigurations(false);
+    }
+
+    @Test(timeout = 300000)
+    public void testConstraintConcurrency() {
+        final String drl =
+                "package com.example.reproducer\n" +
+                           "import " + Bus.class.getCanonicalName() + ";\n" +
+                           "import static " + ConstraintWithAndOrConcurrencyJittingTest.class.getCanonicalName() + ".TOSTRING;\n" +
+                           "dialect \"mvel\"\n" +
+                           "rule \"rule_mt_1a\"\n" +
+                           "    when\n" +
+                           "        $bus : Bus( $check: \"GAMMA RAY\",\n" +
+                           "                    $title: \"POWER PLANT\",\n" +
+                           "                    karaoke.dvd[$title] != null,\n" +
+                           "                    TOSTRING(karaoke.dvd[$title].artist) == \"BBB\" || TOSTRING(karaoke.dvd[$title].artist) >= \"01\" && TOSTRING(karaoke.dvd[$title].artist) <= \"39\" )\n" +
+                           "    then\n" +
+                           "end";
+
+        List<Exception> exceptions = new ArrayList<>();
+
+        //        System.setProperty(MvelEvaluator.THREAD_SAFETY_PROPERTY, "synced_till_eval");
+        System.setProperty(MvelEvaluator.THREAD_SAFETY_PROPERTY, "fully_synced");
+
+        try {
+            for (int i = 0; i < LOOP; i++) {
+                final KieBase kieBase = KieBaseUtil.getKieBaseFromKieModuleFromDrl("accumulate-test", kieBaseTestConfiguration, drl);
+
+                // 1st run : 1 count before jitting
+                for (int jitCount = 0; jitCount < 19; jitCount++) { // if you cannot reproduce the issue, try jitCount < 17 or jitCount < 18
```

Comment:
I can constantly reproduce the issue on my end. If you can't, please try `jitCount < 17` or `jitCount < 18`